device/src/async_device/mod.rs

```diff
@@ -454,105 +457,120 @@ where
         &mut self,
         frame: &Frame,
         window_delay: u32,
+        expected_payload_length: usize,
     ) -> Result<(), Error<R::PhyError>> {
-        let num_read = self.rx_with_timeout_inner(frame, window_delay).await?;
+        let num_read = self.rx_with_timeout_inner(frame, window_delay, expected_payload_length).await?;
         self.radio_buffer.inc(num_read);
         Ok(())
     }
 
+    /// Attempt to receive a packet, waiting at most until `preamble_deadline` to get past a reception of
+    /// a preamble. Then wait at most `message_timeout_ms` to receive the rest of the packet.
+    async fn rx_with_preamble_deadline(
+        &mut self,
+        rx_config: region::RfConfig,
+        preamble_deadline: u32,
+        message_timeout_ms: u32,
+    ) -> Result<(u8, PacketStatus), Error<R::PhyError>> {
+        // Pass the full radio buffer slice to RX
+        self.phy.radio.setup_rx(rx_config).await.map_err(Error::Radio)?;
+        let state = {
+            let rx_preamble_fut = self.phy.radio.rx_until_state(self.radio_buffer.as_raw_slice(), lora_phy::mod_traits::DesiredIrqState::PreambleReceived);
+            let timeout_to_preamble_fut = self.timer.at(preamble_deadline as u64);
+    
+            pin_mut!(rx_preamble_fut);
+            pin_mut!(timeout_to_preamble_fut);
+    
+            match select(rx_preamble_fut, timeout_to_preamble_fut).await {
+                Either::Left((r, _)) => match r {
+                    Ok(state) => {
+                        state
+                    },
+                    Err(err) => {
+                        return Err(Error::Radio(err));
+                    },
+                },
+                Either::Right(_) => return Err(Error::RxTimeout),
+            }
+        };
+
+        match state {
+            IrqState::Waiting => return Err(Error::RxTimeout),
+            IrqState::PreambleReceived => {
+                // continue with reception
+            },
+            IrqState::Done(length, status) => return Ok((length, status)),
+        }
+
+        info!("preamble received, waiting for payload");
+
+        let rx_full_fut = self.phy.radio.rx_until_state(self.radio_buffer.as_raw_slice(), lora_phy::mod_traits::DesiredIrqState::Done);
+        let timeout_fut = self.timer.delay_ms(message_timeout_ms as u64);
+
+        pin_mut!(rx_full_fut);
+        pin_mut!(timeout_fut);
+
+        match select(rx_full_fut, timeout_fut).await {
+            Either::Left((Ok(IrqState::Done(length, lq)), _)) => {
+                Ok((length, lq))
+            },
+            Either::Left((Err(err), _)) => {
+                Err(Error::Radio(err))
+            },
+            _ => {
+                Err(Error::RxTimeout)
+            }
+        }
+    }
+
     async fn rx_with_timeout_inner(
         &mut self,
         frame: &Frame,
         window_delay: u32,
+        expected_payload_length: usize,
```

Comment:
How do you know how many bytes the network will send?

Reply:
ah I see, maybe call it something with maximum

Reply:
renamed into max_expected_payload_length